### PR TITLE
SAM/BAM writers should always call setHeader on alignment records.

### PR DESCRIPTION
**Pull request by cmnbroad**
Includes some DRY-refactoring to support cleaning up mate/reference index state management and javadoc.

### CODE REVIEW

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 1]
```diff
@@ -514,27 +478,79 @@ else if (null != mHeader) {
...
+     *                {@code referenceName} is not NO_ALIGNMENT_REFERENCE_NAME.
+     * @param strict if true, throws if {@code referenceName}  does not appear in the header's sequence dictionary
+     * @returns the reference index corresponding to the {@code referenceName}, or null if strict is false and {@code referenceName}
+     * does not appear in the header's sequence dictionary.
+     * @throws IllegalStateException if {@code referenceName} is not equal to NO_ALIGNMENT_REFERENCE_NAME and the header is null
+     * @throws IllegalArgumentException if strict is true and the name does not appear in header's sequence dictionary.
+     *
+     * Does not mutate the SAMRecord.
+     */
+    protected static Integer resolveIndexFromName(final String referenceName, final SAMFileHeader header, final boolean strict) {
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Please add direct unit tests for both `resolveIndexFromName()` and `resolveNameFromIndex()` that exercise all the possible cases.

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 2]
```diff
@@ -514,27 +478,79 @@ else if (null != mHeader) {
...
+        String referenceName = NO_ALIGNMENT_REFERENCE_NAME;
+        if (NO_ALIGNMENT_REFERENCE_INDEX != referenceIndex) {
+            if (null == header) {
+                throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the reference index or name");
             }
-            else {
-                throw new IllegalArgumentException("Reference index " + mateReferenceIndex + " not found in sequence dictionary.");
+            SAMSequenceRecord samSeq = header.getSequence(referenceIndex);
+            if (null == samSeq) {
+                throw new IllegalArgumentException("Reference name for '" + referenceIndex + "' not found in sequence dictionary.");
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: So there is no valid use case for a "strict" mode here as well to control whether this exception is thrown, correct?

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 3]
```diff
@@ -342,26 +343,19 @@ private boolean hasMateReferenceName() {
...
-            if (-1 != referenceIndex) {
-                setReferenceIndex(referenceIndex); // sets reference name and index
-            }
-            else {
-                mReferenceName = referenceName.intern();
-                mReferenceIndex = null;
-            }
+            mReferenceName = null == mReferenceIndex ?
+                                    referenceName.intern() :
+                                    resolveNameFromIndex(mReferenceIndex, mHeader);
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Why do we make this additional call to `resolveNameFromIndex`() here? Won't this always just give us back the `referenceName` we already had, provided the earlier call to `resolveIndexFromName()` returned non-null?

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 4]
```diff
@@ -342,26 +343,19 @@ private boolean hasMateReferenceName() {
...
+     * @throws IllegalArgumentException if {@code referenceName} is null
      */
     public void setReferenceName(final String referenceName) {
         if (null == referenceName) {
             throw new IllegalArgumentException(
                     "Reference name must not be null. Use SAMRecord.NO_ALIGNMENT_REFERENCE_NAME to reset the reference name.");
         }
-        else if (NO_ALIGNMENT_REFERENCE_NAME.equals(referenceName)) {
-            mReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-            mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Previously if `NO_ALIGNMENT_REFERENCE_NAME` was passed in here, `mReferenceIndex` would get set to `NO_ALIGNMENT_REFERENCE_INDEX` even if the header was null. Now if the header is null, `mReferenceIndex` will get set to null in this case. Is there a reason for this behavioral change? Ideally we shouldn't change any existing behavior when refactoring to use the new `resolve*` methods.

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 5]
```diff
@@ -382,20 +376,14 @@ else if (null != mHeader) {
...
-            }
-            else if (null != mHeader) {
-                mReferenceIndex = mHeader.getSequenceIndex(mReferenceName);
-            }
-            else {
-                throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the reference index");
+        if (null == mReferenceIndex) { // try to resolve the reference index
+            mReferenceIndex = resolveIndexFromName(mReferenceName, mHeader, false);
+            if (null == mReferenceIndex) {
+                return NO_ALIGNMENT_REFERENCE_INDEX;
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Previously this method returned null in this case -- why the change?

Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: This new code leaves `mReferenceIndex` set to null in the case where the lookup failed, whereas the previous version would leave it set to `NO_ALIGNMENT_REFERENCE_INDEX`. Is this what we want?

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 6]
```diff
@@ -410,27 +398,15 @@ else if (null != mHeader) {
...
-                mReferenceIndex = referenceIndex;
-                mReferenceName = samSequence.getSequenceName();
-            }
-            else {
-                throw new IllegalArgumentException("Reference index " + referenceIndex + " not found in sequence dictionary.");
-            }
-        }
+        // resolveNameFromIndex throws if the index can't be resolved
+        setReferenceName(resolveNameFromIndex(referenceIndex, mHeader));
+        mReferenceIndex = referenceIndex;
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Won't `setReferenceName()` set the reference index for us (since we know the header lookup succeeded if the method call even happens)?

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 7]
```diff
@@ -447,24 +423,18 @@ public String getMateReferenceName() {
...
+     * @throws IllegalArgumentException if {@code mateReferenceName} is null
      */
     public void setMateReferenceName(final String mateReferenceName) {
         if (null == mateReferenceName) {
-            throw new IllegalArgumentException("Mate reference name must not be null");
+            throw new IllegalArgumentException("Mate reference name must not be null. Use SAMRecord.NO_ALIGNMENT_REFERENCE_NAME to reset the mate reference name.");
         }
-        else if (NO_ALIGNMENT_REFERENCE_NAME.equals(mateReferenceName)) {
-            mMateReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-            mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Same comment as for `setReferenceName()` above -- the behavior has changed so that if the header is null and `NO_ALIGNMENT_REFERENCE_NAME` is passed in here, `mMateReferenceIndex` ends up as null rather than `NO_ALIGNMENT_REFERENCE_INDEX`. Is this desired?

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 8]
```diff
@@ -447,24 +423,18 @@ public String getMateReferenceName() {
...
-            else {
-                mMateReferenceName = mateReferenceName.intern();
-                mMateReferenceIndex = null;
-            }
+        if (null != mHeader) {
+            mMateReferenceIndex = resolveIndexFromName(mateReferenceName, mHeader, false);
+            // String.intern() is surprisingly expensive, so try to avoid it by looking up in sequence dictionary if possible
+            mMateReferenceName = null == mMateReferenceIndex ?
+                                   mateReferenceName.intern() :
+                                   resolveNameFromIndex(mMateReferenceIndex, mHeader);
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Same comment as for `setReferenceName()` -- why do we make this additional call to resolveNameFromIndex() here?

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 9]
```diff
@@ -486,20 +456,14 @@ else if (null != mHeader) {
...
-            }
-            else if (null != mHeader) {
-                mMateReferenceIndex = mHeader.getSequenceIndex(mMateReferenceName);
-            }
-            else {
-                throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the mate reference index");
+        if (null == mMateReferenceIndex) { // try to resolve the mate reference index
+            mMateReferenceIndex = resolveIndexFromName(mMateReferenceName, mHeader, false);
+            if (null == mMateReferenceIndex) {
+                return NO_ALIGNMENT_REFERENCE_INDEX;
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Same comment as for `getReferenceIndex()` -- This new code leaves mMateReferenceIndex set to null in the case where the lookup failed, whereas the previous version would leave it set to `NO_ALIGNMENT_REFERENCE_INDEX`. Do we care?

File: src/java/htsjdk/samtools/SAMRecord.java [hunk 10]
```diff
@@ -514,27 +478,79 @@ else if (null != mHeader) {
...
     public void setMateReferenceIndex(final int mateReferenceIndex) {
-        if (mateReferenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
-            mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            mMateReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-        }
-        else if (null == mHeader) {
-            throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the mate reference index");
+        // resolveNameFromIndex throws if the index can't be resolved
+        setMateReferenceName(resolveNameFromIndex(mateReferenceIndex, mHeader));
+        mMateReferenceIndex = mateReferenceIndex;
```
Comment (droazen) [inline src/java/htsjdk/samtools/SAMRecord.java]: Same comment as for `setReferenceIndex()` -- Won't setMateReferenceName() set the reference index for us (since we know the header lookup succeeded if resolveNameFromIndex() did not throw)?

File: src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java [hunk 11]
```diff
@@ -847,4 +847,51 @@ private void testNullHeadGetCigarBAM() {
 
         testNullHeaderCigar(bamRec);
     }
+
+    @Test
+    public void testSetHeaderStrictValid() {
+        SAMRecord sam = createTestRecordHelper();
+        final SAMFileHeader samHeader = sam.getHeader();
```
Comment (droazen) [inline src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java]: Add a second `testSetHeaderStrictValid` test case that uses `setHeaderStrict()` to set the header on a read to a header that is compatible, but different from the original header.

Comment (droazen) [inline src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java]: Also test setting a valid header on a headerless record.

File: src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java [hunk 12]
```diff
@@ -847,4 +847,51 @@ private void testNullHeadGetCigarBAM() {
...
+        Assert.assertEquals(sam.getReferenceIndex(), originalRefIndex);
+    }
+
+    @Test(expectedExceptions=IllegalArgumentException.class)
+    public void testSetHeaderStrictInvalidReference() {
+        SAMRecord sam = createTestRecordHelper();
+        final SAMFileHeader samHeader = sam.getHeader();
+
+        sam.setReferenceName("unresolveable");
+        int originalRefIndex = sam.getReferenceIndex();
```
Comment (droazen) [inline src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java]: This is odd -- didn't `setReferenceName()` clear the reference index? (and so "originalRefIndex" is not actually the "original" reference index)

**Comment by droazen** [comment]
First-pass review complete -- back to @cmnbroad

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 3 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Its just to maintain the interning pattern that was already in place before, and retrieve the already interned object. Not sure how much its buying us. I must make that comment more explicit because even with it I had to think about it a bit to recall why I did that.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 6 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Yes, I think I was just trying to be explicit about it since its the main point of the call.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 4 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Good point.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 5 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Also a good point.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 7 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: No. Needs fixing.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 8 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Interning.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 9 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Also needs fixing.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 10 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Just being explicit.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 1 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Done.

File: src/java/htsjdk/samtools/SAMRecord.java (hunk 2 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/SAMRecord.java]: Right - I originally had a strict mode for this as well but the policy was and is that you can't set an unresolveable index on a record, only an unresolveable name.

File: src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java (hunk 11 above)
Comment (cmnbroad) [inline src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java]: Both done.

File: src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java (hunk 12 above)
Comment (cmnbroad) [inline src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java]: Right, done.

**Comment by cmnbroad** [comment]
Back to @droazen.

Also, a while back we agreed that you should be able to use setHeaderStrict to set a null header and just mirror the behavior of setHeader for that case. So I made that change in this round, which affected one test.

File: src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java [hunk 13]
```diff
@@ -847,4 +847,138 @@ private void testNullHeadGetCigarBAM() {
...
+    public void testResolveIndexResolveable() {
+        final SAMRecord sam = createTestRecordHelper();
+        final SAMFileHeader samHeader = sam.getHeader();
+        final String contigName = sam.getContig();
+        Assert.assertEquals(SAMRecord.resolveIndexFromName(contigName, samHeader, true), new Integer(samHeader.getSequenceIndex(contigName)));
+    }
+
+    @Test(expectedExceptions=IllegalStateException.class)
+    public void testResolveIndexUnresolveableStrict() {
+        Assert.assertEquals(SAMRecord.resolveIndexFromName("unresolveable", null, false), new Integer(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
```
Comment (droazen) [inline src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java]: Test case is mis-named (strict mode is off here). This is testing whether you throw when the header is null and strict mode is off. You also need a separate test with an unresolvable contig, a non-null header, and strict mode on that expects an exception.

**Comment by droazen** [comment]
:+1: Branch looks good now. One final TODO in the tests, then please merge this yourself. Back to @cmnbroad
